iree/tools/iree-run-mlir-main.cc

```diff
@@ -56,19 +64,34 @@
 #include "iree/tools/init_targets.h"
 #include "iree/tools/utils/vm_util.h"
 #include "iree/vm/api.h"
-#include "iree/vm/bytecode_module.h"
+#include "iree/vm/ref_cc.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/ADT/iterator.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Casting.h"
 #include "llvm/Support/CommandLine.h"
+#include "llvm/Support/ErrorHandling.h"
 #include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SMLoc.h"
 #include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/raw_ostream.h"
 #include "mlir/IR/AsmState.h"
-#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BlockSupport.h"
+#include "mlir/IR/BuiltinAttributes.h"
 #include "mlir/IR/BuiltinOps.h"
 #include "mlir/IR/Dialect.h"
 #include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/OwningOpRef.h"
```

Comment:
That's a lot of includes :O

Wonder if we'd want to respect LLVM's style (https://llvm.org/docs/CodingStandards.html#minimal-list-of-includes) where it becomes noisy like here

Reply:
We do respect it - it states `You must include all of the header files that you are using` :)

The note they have about includes slowing down compilation is not true in any modern compiler - if the header is required then it's going to be included somewhere, and compilers are exceedingly good at caching headers with #ifndef/#pragma once guards. I find the LLVM style of "guess what you need and hope someone else doesn't change their includes and break you" quite sloppy :)

Even if not using iwyu adding the `export` comment is a good way to denote that a header exposes a transitive inclusion as part of its API - otherwise it's just incidental if you get things. For example, if some class has a custom ostream override and includes `raw_ostream.h` and you rely on that and one day someone moves that to a util file your code will break (unless some _other_ include you have happens to be including it -- for now). MLIR commits similar sins by including _some_ of ADT and _some_ forward declarations so you never know which ADT files you should include (no SmallVector, yes SmallSet, etc) 🤮

Reply:
(and yeah, it's a lot of includes - would be nice if there were some wrappers here to at least localize them, but this whole file should probably be rewritten when we have a C compiler API and don't leak any LLVM headers :)